src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -666,5 +581,241 @@ void BaseQNAcceleration::writeInfo(
+
+        for (Eigen::Index i = 0; i < dataSize; i++) {
+          primaryValues(i + offset)    = primaryData(i);
+          oldPrimaryValues(i + offset) = oldPrimaryData(i);
+        }
+        offset += dataSize;
+      }
+    }
+  } else {
+    Eigen::Index offset = 0;
+    for (auto id : _primaryDataIDs) {
+      Eigen::Index dataSize       = cplData.at(id)->values().size();
+      auto &       primaryData    = cplData.at(id)->values();
+      const auto & oldPrimaryData = cplData.at(id)->previousIteration();
+
+      PRECICE_ASSERT(primaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+      PRECICE_ASSERT(oldPrimaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+
+      for (Eigen::Index i = 0; i < dataSize; i++) {
+        primaryValues(i + offset)    = primaryData(i);
+        oldPrimaryValues(i + offset) = oldPrimaryData(i);
+      }
+      offset += dataSize;
+    }
+  }
+  /// Sample all the data to the corresponding time grid in _timeGrids and concatenate everything into a long vector
+  Eigen::Index offset = 0;
+
+  for (int id : _dataIDs) {
+    auto            waveform = cplData.at(id)->timeStepsStorage();
+    Eigen::Index    dataSize = cplData.at(id)->values().size();
+    Eigen::VectorXd timeGrid = _timeGrids.at(id);
+
+    for (int i = 0; i < timeGrid.size(); i++) {
+
+      Eigen::VectorXd data    = waveform.sample(timeGrid(i));
+      Eigen::VectorXd oldData = cplData.at(id)->getPreviousValuesAtTime(timeGrid(i));
+
+      PRECICE_ASSERT(values.size() >= offset + dataSize, "the values were not initialized correctly");
+
+      for (Eigen::Index i = 0; i < dataSize; i++) {
+        values(i + offset)    = data(i);
+        oldValues(i + offset) = data(i);
+      }
+      offset += dataSize;
+    }
+  }
+}
+
+void BaseQNAcceleration::saveTimeGrid(const DataMap &cplData)
+{
+  for (auto &pair : cplData) {
+    auto            dataID   = pair.first;
+    Eigen::VectorXd timeGrid = pair.second->timeStepsStorage().getTimes();
+    _timeGrids.insert(std::pair<int, Eigen::VectorXd>(dataID, timeGrid));
+  }
+}
+
+void BaseQNAcceleration::initializeVectorsAndPreconditioner(const DataMap &cplData, const std::vector<DataID> &dataIDs, const std::vector<DataID> &primaryDataIDs)
+{
+
+  size_t dataSize = 0;
+  for (auto id : dataIDs) {
+    dataSize += cplData.at(id)->timeStepsStorage().nTimes() * cplData.at(id)->values().size();
+  }
+
+  size_t              primaryDataSize = 0;
+  std::vector<size_t> subVectorSizes; // needed for preconditioner
+
+  for (auto id : primaryDataIDs) {
+    if (_reduced) {
+      primaryDataSize += cplData.at(id)->values().size();
+      subVectorSizes.push_back(cplData.at(id)->getSize());
+    } else {
+      primaryDataSize += cplData.at(id)->timeStepsStorage().nTimes() * cplData.at(id)->values().size();
+      subVectorSizes.push_back(cplData.at(id)->timeStepsStorage().nTimes() * cplData.at(id)->getSize());
```

Comment:
```suggestion
      primaryDataSize += cplData.at(id)->getSize();
      subVectorSizes.push_back(cplData.at(id)->getSize());
    } else {
      primaryDataSize += cplData.at(id)->timeStepsStorage().nTimes() * cplData.at(id)->getSize();
      subVectorSizes.push_back(cplData.at(id)->timeStepsStorage().nTimes() * cplData.at(id)->getSize());
```
`values().size()` and getSize()` is identical. I would avoid using `values().size()` (see https://en.wikipedia.org/wiki/Law_of_Demeter).